File: hotel-app/src/main/java/eu/getsoftware/hotelico/service/booking/config/adapter/kafka/KafkaProducerConfig.java
package eu.getsoftware.hotelico.service.booking.config.adapter.kafka;

import eu.getsoftware.hotelico.clients.api.application.infrastructure.domainevents.domainmessage.DomainMessage;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.core.ProducerFactory;
import org.springframework.kafka.support.serializer.JsonSerializer;

import java.util.HashMap;
import java.util.Map;

@Configuration
@ConditionalOnProperty(name = "app.kafka.enabled", havingValue = "true", matchIfMissing = true)
public class KafkaProducerConfig {
    
    @Value("${spring.kafka.bootstrap-servers}")
    private String bootstrapServers;
    
    public Map<String, Object> producerConfig() {
        Map<String, Object> props = new HashMap<>();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class);
        return props;
    }
    
    @Bean
    public ProducerFactory<String, DomainMessage> producerFactory() {
        return new DefaultKafkaProducerFactory<>(producerConfig());
    }

    /**
     * eu: create KafkaTemplate on basis of ProducerFactory
     * @param producerFactory
     * @return
     */
    @Bean
    public KafkaTemplate<String, DomainMessage> kafkaTemplate(
            ProducerFactory<String, DomainMessage> producerFactory
    ) {
        return new KafkaTemplate<>(producerFactory);
    }
}

File: hotel-service-chat/src/main/kotlin/eu/getsoftware/hotelico/chat/config/adapter/kafka/KafkaProducerConfig.kt
package eu.getsoftware.hotelico.chat.config.adapter.kafka

import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.protocol.Message
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import org.springframework.kafka.support.serializer.JsonSerializer

@Configuration
class KafkaProducerConfig {
    @Value("\${spring.kafka.bootstrap-servers}")
    private val bootstrapServers: String? = null

    fun producerConfig(): Map<String, Any?> {
        val props: MutableMap<String, Any?> = HashMap()
        props[ProducerConfig.BOOTSTRAP_SERVERS_CONFIG] = bootstrapServers
        props[ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG] =
            StringSerializer::class.java
        props[ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG] =
            JsonSerializer::class.java
        return props
    }

    @Bean
    fun producerFactory(): ProducerFactory<String, Message> {
        return DefaultKafkaProducerFactory(producerConfig())
    }

    /**
     * eu: create KafkaTemplate on basis of ProducerFactory
     * @param producerFactory
     * @return
     */
    @Bean
    fun kafkaTemplate(
        producerFactory: ProducerFactory<String, Message>
    ): KafkaTemplate<String, Message> {
        return KafkaTemplate(producerFactory)
    }
}
